src/utils/__tests__/resolve_url.test.ts

```diff
@@ -16,28 +16,31 @@
 
 import resolveURL, { getFilenameIndexInUrl } from "../resolve_url";
 
-describe("utils - resolveURL", () => {
+describe(`utils - resolveURL ${resolveURL.name}`, () => {
   it("should return an empty string if no argument is given", () => {
     expect(resolveURL()).toBe("");
   });
 
   it("should concatenate multiple URLs", () => {
     expect(resolveURL("http://toto.com/a")).toBe("http://toto.com/a");
-    expect(resolveURL("http://toto.com/a", "b/c/d/", "g.a")).toBe(
-      "http://toto.com/a/b/c/d/g.a",
-    );
+    expect(
+      resolveURL("http://toto.com/a" /** Trailling slash missing */, "b/c/d/", "g.a"),
+    ).toBe("http://toto.com/b/c/d/g.a");
+    expect(
+      resolveURL("http://toto.com/a/" /** With a trailling slash */, "b/c/d/", "g.a"),
```

Comment:
Just to clarify why I'm wondering for this one, if the URL given to the loadVideo is something like:

```js
player.loadVideo({ url: "https://example.com/my-mpd.mpd/" })
```

And if there's a `media="toto"` inside the MPD but not `<BaseURL>`, should the URL be:
`"https://example.com/my-mpd.mpd/toto"` (as is implied here)
or
``"https://example.com/toto"`

From the perspective of only the DASH specification (which may or may not defer this URL resolution entirely to the RFC3986)?

TBH the most probable is that you're right but I want to double check to be sure here!

Reply:
My understanding is that the URL at which the manifest is downloaded is NOT considered to be a baseURL, as opposed to elements `<BaseURL>` 

From the ISO 23009-1 describing DASH, it says in 5.6.4: 

> URLs at each level of the MPD are resolved according to IETF RFC 3986 with respect to the BaseURL element specified at that level of the document, or the level above in the case of resolve base URLs themselves (the document "base URI" as defined in IETF RFC 3986;2005, subclause 5.1 is considered to be the level above the MPD level). If only relative URLs are specified and the document base URI cannot be established according to IETF RFC 3986 then the MPD should not be interpreted

If manifest URL is `https://example.com/my-mpd.mpd/` and there is no `<BaseURL>` in the MPD,  by default, base URL is the level above the MPD level: in it's case base URL should be `https://example.com/`. 
Although it's not explicit what "level above" mean.

It's also said that:
> NOTE URL resolution is defined in RFC 3986:2005 section 5.2 and is not a simple concatenation. In particular, when a URL in the MPD is specified as a relative URLs, then as per RFC 3986 for the URL resulting from resolving the value of the BaseURL element, all characters following the last "/" part of its path part are replaced by the value of the actual URL attribute

---
Example 1: No `<BaseURL> with a trailing slash`

```xml
<! -- manifest downloaded from https://example.com/my-mpd.mpd/ -->
<MPD>
  <SegmentTemplate media="toto">
</MPD>
```
No `<BaseURL>` => the baseURL is the document base URI (which is defined as the level above the MPD level). 

**baseURL**: https://example.com/ (level above MPD)
**relativeURL**: toto
**resolvedURL**: https://example.com/toto


---
Example 2: No `<BaseURL> without a trailing slash`

```xml
<! -- manifest downloaded from https://example.com/my-mpd.mpd -->
<MPD>
  <SegmentTemplate media="toto">
</MPD>
```
No `<BaseURL>` => the baseURL is the document base URI (which is defined as the level above the MPD level). 

**baseURL**: https://example.com/ (level above MPD - the trailing slash does not matter here)
**relativeURL**: toto
**resolvedURL**: https://example.com/toto

--- 

Example 3: `<BaseURL> with a trailing slash`

```xml
<! -- manifest downloaded from https://example.com/my-mpd.mpd/ -->
<MPD>
  <BaseURL>https://example.com/my-mpd.mpd/<BaseURL>
  <SegmentTemplate media="toto">
</MPD>
```
found a `<BaseURL>`, remove all characters following the last "/" part and use it as is. Here there is nothing after last "/" so there is nothing to remove.
**baseURL**: https://example.com/my-mpd.mpd/
**relativeURL**: toto
**resolvedURL**: https://example.com/my-mpd.mpd/toto

--- 


--- 

Example 4: `<BaseURL> with a trailing slash`

```xml
<! -- manifest downloaded from https://example.com/my-mpd.mpd/ -->
<MPD>
  <BaseURL>https://example.com/my-mpd.mpd<BaseURL>
  <SegmentTemplate media="toto">
</MPD>
```

found a `<BaseURL>`, remove all characters following the last "/" part and use it as is. Here it removes "toto"
**baseURL**: https://example.com/
**relativeURL**: toto
**resolvedURL**: https://example.com/my-mpd.mpd/toto

--- 


If there is a server that redirect `https://example.com/my-mpd.mpd/` to `https://example.com/my-mpd.mpd` and the player was loaded with ```player.loadVideo({ url: "https://example.com/my-mpd.mpd/" })``` 
I doubt it will work, maybe we will need to update how the "initialBaseURL" is obtain to not have that trailing slash ?

```js
// src/parsers/manifest/dash/common/parse_mpd.ts
  const initialBaseUrl: IResolvedBaseUrl[] =
    args.url !== undefined
      ? [{ url: args.url.substring(0, getFilenameIndexInUrl(args.url)) }]
      : [];

```

Reply:
To me, the following line seems a little ambiguous:

> or the level above in the case of resolve base URLs themselves (the document "base URI" as defined in IETF RFC 3986;2005, subclause 5.1 is considered to be the level above the MPD level)

What does "level above" mean in that exact case (example 1) - or perhaps a trailing slash isn't legal here?

To me we should check that we're aligned with other players at worst for that particular scenario.

---

As for HTTP redirect in general, I think we do consider the MPD URL post-redirect, to check. What I know is that some applications expect that at Canal+

Is that aligned with the spec?

---

Just so we're clear, I'm not talking about someone wrongly typing `"https://example.com/my-mpd.mpd/"`.
I'm wondering what should be the resolution for that scenario without considering potential typos here, just from the point of view of the specification and the expected behavior.

Reply:
I have tested with a redirect with status code 301: `"https://example.com/my-mpd.mpd/"` => `"https://example.com/my-mpd.mpd"`. 
The video is played normally and segment request are made to `"https://example.com/segment.mp4"`.

In fact we consider the "real" URL (after a redirect) and not the original URL in `src/transports/dash/manifest_parser.ts`, so the behavior seems good here. 
```js
 const url = manifestData.url ?? parserOptions.originalUrl;
```
Other player such as Shaka or DashJS do the same and are able to play the content.

----
I have made another test with "https://example.com/my-mpd.mpd/" that serves directly the file,
segments are requested to 
"https://example.com/my-mpd.mpd/segment.mp4" and it fails with a 404 because such a file does not exist.
Shaka and DashJS do the same and they both fail to play content. So I think it's normal to fail here.


To sum-up: 
In case of an URL with a trailing slash,
if the server do proper redirect, it's working. 
If the server serves directly the manifest, it's failing.

Reply:
OK that's perfect and there's no issue then.

Sorry for that supplementary step, I wanted to profit from this to make sure that we handled every corner of the specs as wanted before it's part of the next release.